cloud/linode/instances_test.go

```diff
@@ -245,6 +245,89 @@ func TestMalformedProviders(t *testing.T) {
 	})
 }
 
+func TestProviderByIP(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	client := NewMockClient(ctrl)
+
+	t.Run("gets linode by external IP", func(t *testing.T) {
+		instances := newInstances(client)
+
+		publicIP := net.ParseIP("172.234.31.123")
+		privateIP := net.ParseIP("192.168.159.135")
+		wrongIP := net.ParseIP("1.2.3.4")
+		expectedInstance := linodego.Instance{ID: 12345, IPv4: []*net.IP{&publicIP, &privateIP}}
+		instances.nodeCache.nodes = map[int]*linodego.Instance{0: {ID: 54363, IPv4: []*net.IP{&wrongIP}}, 1: &expectedInstance}
+		node := v1.Node{Status: v1.NodeStatus{Addresses: []v1.NodeAddress{
+			{
+				Type:    "Hostname",
+				Address: "node1.linodelke.com",
+			}, {
+				Type:    "ExternalIP",
+				Address: "172.234.31.123",
+			},
+		}}}
+		k8sNode, err := instances.linodeByIP(&node)
+		assert.Nil(t, err)
+		assert.Equal(t, expectedInstance.ID, k8sNode.ID)
+	})
+
+	t.Run("gets linode by internal IP", func(t *testing.T) {
+		instances := newInstances(client)
+
+		publicIP := net.ParseIP("172.234.31.123")
+		privateIP := net.ParseIP("192.168.159.135")
+		wrongIP := net.ParseIP("1.2.3.4")
+		expectedInstance := linodego.Instance{ID: 12345, IPv4: []*net.IP{&publicIP, &privateIP}}
+		instances.nodeCache.nodes = map[int]*linodego.Instance{0: {ID: 54363, IPv4: []*net.IP{&wrongIP}}, 1: &expectedInstance}
```

Comment:
You're leveraging an implementation detail for this test to work. I'd prefer if you didn't populate the cache to force a result, we have better ways of doing this - there's a mock for the API (and there's even a fake API implementation), but the former is more than sufficient for this kind of test.

Reply:
I've refactored the tests to use public methods instead of using implementation details with the private methods